src/swarm-lang/Swarm/Language/Types.hs

```diff
@@ -650,6 +728,34 @@ type TCtx = Ctx Polytype
 --   are in the midst of the type inference process.
 type UCtx = Ctx UPolytype
 
+-- | A @TVCtx@ tracks which type variables are in scope, and what
+--   skolem variables were assigned to them.
+type TVCtx = Ctx UType
+
+------------------------------------------------------------
+-- Implicit quantification of polytypes
+------------------------------------------------------------
+
+-- | Implicitly quantify any otherwise unbound type variables.
+quantify ::
+  (Has (Reader TVCtx) sig m, Typical ty) =>
+  Poly 'Unquantified ty ->
+  m (Poly 'Quantified ty)
+quantify (Forall xs ty) = do
+  -- Look at all variables which occur in the type but are not explicitly bound by the forall...
+  let unbound = tyVars ty `S.difference` S.fromList xs
+  -- ...and filter out those that are not bound in the context and
+  -- should therefore be implicitly quantified
+  inScope <- ask @TVCtx
+  let implicit = S.filter (not . (`Ctx.member` inScope)) unbound
```

Comment:
The `unCtx` is published so you could use another `S.difference` with its `keysSet`. 
I am unsure if it would be efficient, but it could be a nice oneliner, and the comment would not be broken up.